src/v/utils/object_pool.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+
+#include <concepts>
+#include <deque>
+#include <stack>
+
+template<std::movable T>
+class object_pool {
+    struct wait_item {
+        ss::promise<T> p;
+        ss::abort_source::subscription abort_sub;
+    };
+
+public:
+    ss::future<T> allocate_object(ss::abort_source& as) {
+        // Return any available object right away without setting up
+        // a waiter.
+        if (!_objects.empty()) {
+            auto ret = ss::make_ready_future<T>(std::move(_objects.top()));
+            _objects.pop();
+            return ret;
+        }
+
+        auto item = std::make_unique<wait_item>();
+
+        auto sub_opt = as.subscribe([this, item_ptr = item.get()]() noexcept {
+            // it should be safe to use a raw pointer as the lifetime of this callback
+            // is the lifetime of the allocation the pointer points to.
+            item_ptr->p.set_exception(ss::abort_requested_exception());
+            std::erase_if(
+              _waiters, [item_ptr](const std::unique_ptr<wait_item>& it) {
+                  return it.get() == item_ptr;
+              });
+        });
+
+        if (!sub_opt) {
+            // Abort source already fired!
+            return ss::make_exception_future<T>(
+              ss::abort_requested_exception());
+        }
+
+        item->abort_sub = std::move(*sub_opt);
+        _waiters.emplace_back(std::move(item));
+        return _waiters.back()->p.get_future();
+    }
+
+    void release_object(T obj) {
+        if (!_waiters.empty()) {
+            _waiters.front()->p.set_value(std::move(obj));
+            _waiters.pop_front();
+        } else {
+            _objects.emplace(std::move(obj));
+        }
+    }
+
+private:
+    // try to perserve any locality by reusing the
+    // most recently used object
+    std::stack<T> _objects;
+    std::deque<std::unique_ptr<wait_item>> _waiters;
```

Comment:
for seastar this is not as general of an object pool in that most object pools we have have a timer-reclaimer that frees unused buffers w/ timers, usually every 1 second or something like that to be used by other pools. to me a built-in gc may make this class more useful generically. 

also /utils doesn't seem like a good namespace. something like `/data-types` or something like that fore more generic types. what absl does is you import every data type directly. i.e.: `absl::hash_map` as a library. so it maybe worth creating a `adt/object-pool/{cmake, .h, .cc, .test.cc` kinda structure for future types.

Reply:
Reclaiming unused objects that haven't been allocated outside of the object pool would be rather easy. I.e, objects just sitting in the `_objects` stack in the class above.

A lot of our other object pools like the `batch_cache` allow for reclaiming objects that have already been allocated though. From what I can tell the way they achieve this is by handing out objects wrapped in some class. That wrapper class records how often a user access the underlying object. Then the object pool deallocates objects a user hasn't accessed in awhile.

The issue I can see with generalizing this pattern is that a user would have to be sure to check that the wrapped object is still available every time they want to access it. They would also have to check that it's still available across seastar scheduling points as well as it could've been deallocated during a scheduling point. It just seems a bit error-prone.

Reply:
In fact, the batch cache is the _only_ class that allows this kind of async reclaim driven by memory pressure, right? At least, it's the only class registered as a "reclaimer" with the seastar allocator: the allocator only lets you register one anyway, so if you want multiple you'd have to write some kind of meta-reclaimer which holds a list of child reclaimers and calls them based on some rules when seastar says "give me more memory".

Reply:
@travisdowns i think seastar supports multi-reclaimers in a first-registered priority order. the fun part is that a new reclaimer is registered in seastar merely by constructing a reclaimer object.

Reply:
@dotnwat oh interesting, I had looked at the API briefly and thought it was only 1 reclaim system-wide (well, shard-wide) but I missed the implicit creation of a registered reclaimer I guess.

Reply:
i think in practice you'd want to register one and implement a custom priority